client/src/procedures/primitives.rs

```diff
@@ -0,0 +1,1008 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{enum_from_inner, Location};
+
+use super::*;
+use crypto::{
+    ciphers::{
+        aes::Aes256Gcm,
+        chacha::XChaCha20Poly1305,
+        traits::{consts::Unsigned, Aead, Tag},
+    },
+    hashes::{
+        blake2b::Blake2b256,
+        sha::{Sha256, Sha384, Sha512},
+        Digest,
+    },
+    keys::{
+        bip39,
+        slip10::{self, Chain, ChainCode, Curve, Seed},
+        x25519,
+    },
+    signatures::ed25519,
+    utils::rand::fill,
+};
+use engine::{runtime::GuardedVec, vault::RecordHint};
+use hmac::{
+    digest::{BlockInput, FixedOutput, Reset, Update},
+    Mac, NewMac,
+};
+use serde::{Deserialize, Serialize};
+use std::{convert::TryFrom, marker::PhantomData};
+use stronghold_derive::{execute_procedure, Procedure};
+
+// ==========================
+// Helper Procedures
+// ==========================
+
+#[derive(Clone, GuardDebug, Serialize, Deserialize)]
+pub enum PrimitiveProcedure {
+    Helper(HelperProcedure),
+    Crypto(CryptoProcedure),
+}
+
+impl ProcedureStep for PrimitiveProcedure {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error> {
+        match self {
+            PrimitiveProcedure::Helper(proc) => proc.execute(runner, state),
+            PrimitiveProcedure::Crypto(proc) => proc.execute(runner, state),
+        }
+    }
+}
+
+#[derive(Clone, GuardDebug, Serialize, Deserialize)]
+pub enum HelperProcedure {
+    WriteVault(WriteVault),
+}
+
+impl ProcedureStep for HelperProcedure {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error> {
+        match self {
+            HelperProcedure::WriteVault(proc) => proc.execute(runner, state),
+        }
+    }
+}
+
+// TODO: remove notes, add proper docs.
+#[derive(Clone, GuardDebug, Serialize, Deserialize)]
```

Comment:
```suggestion
#[derive(Clone, GuardDebug, Serialize, Deserialize)]
#[non_exhaustive]
```

To future-proof the APi, better add `#[non_exhaustive]` to the enum, so added variants might be handled with a proper error.